homeassistant/components/sensor/bloomsky.py

```diff
@@ -0,0 +1,104 @@
+"""
+homeassistant.components.sensor.bloomsky
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Support for BloomSky weather station.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/bloomsky/
+"""
+import logging
+import homeassistant.components.bloomsky as bloomsky
+from homeassistant.helpers.entity import Entity
+
+DEPENDENCIES = ["bloomsky"]
+
+# these are the available sensors
+SENSOR_TYPES = ["Temperature",
+                "Humidity",
+                "Rain",
+                "Pressure",
+                "Luminance",
+                "Night",
+                "UVIndex"]
+
+# sensor units - these do not currently align with the API documentation
+SENSOR_UNITS = {"Temperature": "°F",
+                "Humidity": "%",
+                "Pressure": "inHg",
+                "Luminance": "cd/m²"}
+
+# which sensors to format numerically
+FORMAT_NUMBERS = ["Temperature", "Pressure"]
+
+
+# pylint: disable=unused-argument
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """ Set up the available BloomSky weather sensors """
+
+    logger = logging.getLogger(__name__)
+
+    for device_key in bloomsky.BLOOMSKY.devices:
+        device = bloomsky.BLOOMSKY.devices[device_key]
+        for variable in config["monitored_conditions"]:
+            if variable in SENSOR_TYPES:
+                add_devices([BloomSkySensor(bloomsky.BLOOMSKY,
+                                            device,
+                                            variable)])
+            else:
+                logger.error("Cannot find definition for device: %s", variable)
+
+
+class BloomSkySensor(Entity):
+    """ Represents a single sensor in a BloomSky device """
+
+    def __init__(self, bs, device, sensor_name):
+        self._bloomsky = bs
+        self._device_id = device["DeviceID"]
+        self._client_name = device["DeviceName"]
+        self._sensor_name = sensor_name
+        self._state = self.process_state(device)
+        self._sensor_update = ""
+
+    @property
+    def name(self):
+        """ the name of the BloomSky device and this sensor """
+        return "{} {}".format(self._client_name, self._sensor_name)
+
+    @property
+    def state(self):
+        """ the current state (i.e. value) of this sensor """
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """ this sensor's units """
+        return SENSOR_UNITS.get(self._sensor_name, None)
+
+    def update(self):
+        """ request an update from the BloomSky API """
+        self._bloomsky.refresh_devices()
+        # TS is a Unix epoch timestamp for the last time the BloomSky servers
+        # heard from this device. If that value hasn't changed, the value has
+        # not been updated.
+        last_ts = self._bloomsky.devices[self._device_id]["Data"]["TS"]
+        if last_ts != self._sensor_update:
+            self.process_state(self._bloomsky.devices[self._device_id])
```

Comment:
I broke that out into its own method because it's called both in the `update` method as well as the `__init__` constructor (the constructor already has the device info when it is called).